doc/developers/maintainer.rst.template

```diff
@@ -353,6 +356,46 @@ following script and enter the token when prompted:
   cd build_tools
   make authors  # Enter the token when prompted
 
+.. _bumping_dependencies_guideline:
+
+Guideline for bumping minimum versions of our dependencies
+----------------------------------------------------------
+
+- **minimum Python version**: at the time of a minor scikit-learn release (`X.Y.0`),
+  we drop Python whose initial release was more than 4 years old. In other words,
+  our minimum Python version is between 3 and 4 years old.
+- **compiled dependencies** (numpy, scipy, as well as compiled optional
+  dependencies (pandas, matplotlib, pyamg, pillow, ...): we take the oldest minor
+  release (`X.Y.0`) that has wheels for our minimum Python version. In practice
+  this means that our minimum supported version is around 3 years old, maybe a
+  bit less.
+- pure Python dependencies (joblib, threadpoolctl): at the time of the
+  scikit-learn release our minimum supported version is the most recent minor
+  release (`X.Y.0`) that is at least 2 years old.
+- we may decide to be less conservative than this guideline in some edge cases.
+  These edge cases include: a security bugfix in one of our dependencies or a
+  critical bugfix in one of our dependencies makes it too costly to support it in
+  terms of maintenance.
+
+
+Example for the upcoming scikit-learn 1.8 release
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The upcoming 1.8 release is scheduled around November/December, let's pretend
+it will happen on 2025-12-01, Python 3.10 (released in October 2021) will be
+more than 4 years old and will be dropped so our minimum supported Python will
+be Python 3.11 (a bit more than 3 years old).
+
+The dependencies minimum versions will be bumped as follow:
+
+- python: 3.10 -> 3.11
+- numpy: 1.22.0 -> 1.24.0 (oldest version with Python 3.11 wheels)
+- scipy: 1.8.0 -> 1.10.0 (oldest scipy version with Python 3.11 wheels)
+- pandas: 1.4.0 -> 1.5.0 (oldest pandas version with Python 3.11 wheels)
+- matplotlib: 3.5.0 -> 3.6.0 (oldest matplotlib version with Python 3.11 wheels)
+- joblib: 1.2.0 -> 1.3.0 (1.3.0 will be ~2 years 5 month old, 1.4.0 will be less than 2 years old)
+- threadpoolctl: 3.1.0 -> 3.2.0 (3.2.0 will be ~2 years 5 months old, 3.3.0 will be less than 2 years old)
+
```

Comment:
I don't think this part is needed.

Instead we could consolidate and include the script you linked in https://github.com/scikit-learn/scikit-learn/pull/30895#discussion_r2015732106 in `maint_tools` and link that here :)

Reply:
I find an example helps understanding the rule but it also makes sense to keep it short. I guess I can always add a link to the github comment instead.

Reply:
I find that it quickly becomes out-of-date and we let very old examples with very old version numbers in the doc that we never update or clean-up.
At least we should rephrase it a bit such that it's not using time relative to now and use absolute timings instead. For instance `the upcoming scikit-learn 1.8 release`. Similarly, `The dependencies minimum versions will be bumped as follow:` should be `The dependencies minimum versions should be bumped as follow:`.

Reply:
I ended up adding the slightly hacky script I currently have. It can certainly be improved in future PRs.